### DISCUSSION TITLE
Wildcard Tracker Adding

### DISCUSSION THREAD

**Discussion by SalmonSays** [fedarovich/qbittorrent-cli#64] [Q&A] [answered]
Is there no way to add a tracker to every single torrent? 

Preface- The v4.4.3.1 build of qBittorrent (and probably all others, but just for reference I'm up to date) has a web interface that recreates some of the features of the client version, including adding/editing trackers. Instead of right clicking on the torrent itself, the field is moved down to the bottom strip of the web interface. Unlike the client however, selecting all torrents and editing to add a new field only alters the first selected torrent. On the desktop variant, adding a new field updates all existing or selected torrents when multiple are highlighted. In the absence of such a web feature, I was hoping to use this CLI tool to fix what needs fixing on my instance without access to a client since this is remote and dockerized.

Using the qBittorrent-CLI tool I've been able to access my web interface successfully, but using "torrent tracker add https://github.com/ngosang/trackerslist/" tells me the URL field is required because it assumes the first argument is the HASH field. I've also tried "torrent tracker add * https://github.com/ngosang/trackerslist/" to wildcard all torrents, but no dice. Is there a correct way to do this already beyond the methods I've tried which would work?

**Comment by fedarovich** [comment, marked as answer]
Well, while qBittorrent-CLI relies on the same API as used by the web interface, it does not support it directly.

However, you can do it quite easily using the features of your command line shell.

For `bash` you can use the following command:
```bash
qbt torrent list -F json | jq .[].hash | xargs -n 1 -I {} qbt torrent tracker add {} https://raw.githubusercontent.com/ngosang/trackerslist/master/trackers_best.txt
```

and for `powershell` this one:
```powershell
(qbt torrent list -F json | ConvertFrom-JSON).hash | %{& qbt torrent tracker add $_  https://raw.githubusercontent.com/ngosang/trackerslist/master/trackers_best.txt}
```

Both these scripts perform the same steps:

1. Get the torrent list in JSON format.
2. Extract torrent hashes from it (for `bash` we use `jq` utility that should be installed if not present)
3. Invoke `qbt torrent tracker add` for each of the hashes.

**Reply by SalmonSays** [reply]
Presumably "ConvertFrom-JSON" on Powershell is a placeholder for a local utility?

**Reply by fedarovich** [reply]
[ConvertFrom-JSON](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-json?view=powershell-7.2) is a standard Cmdlet included in PowerShell since version 3.0
